service/src/main/java/com/epam/esm/services/service/impl/CertificateServiceImpl.java

```diff
@@ -0,0 +1,197 @@
+import com.epam.esm.services.service.CertificateService;
+import com.epam.esm.services.validator.EntityValidator;
+import org.apache.commons.lang3.StringUtils;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.math.BigDecimal;
+import java.net.URLDecoder;
+import java.nio.charset.StandardCharsets;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+@Service
+public class CertificateServiceImpl implements CertificateService {
+
+    private CertificateDAOImpl dao;
+    private EntityValidator<Certificate> validator;
+    private CertificateFinder finder;
+    private TagServiceImpl tagServiceImpl;
+
+    @Autowired
+    public CertificateServiceImpl(CertificateDAOImpl dao,
+                                  EntityValidator<Certificate> validator,
+                                  CertificateFinder finder, TagServiceImpl tagServiceImpl) {
+        this.dao = dao;
+        this.validator = validator;
+        this.finder = finder;
+        this.tagServiceImpl = tagServiceImpl;
+    }
+
+    @Override
+    public Certificate create(Certificate certificate) throws ServiceException, ValidationException {
+        try {
+            validator.validate(certificate);
+            return dao.create(certificate);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public Certificate read(int id) throws ServiceException {
+        try {
+            return dao.read(id);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public void delete(int id) throws ServiceException {
+        try {
+            if (dao.read(id) == null) {
+                throw new ServiceException("Entity does not exist!");
+            }
+            dao.delete(id);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public void update(Certificate certificate) throws ServiceException, ValidationException {
+        try {
+            validator.validate(certificate);
+            dao.update(certificate);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public List<Certificate> findAll() throws ServiceException {
+        try {
+            return dao.findAll();
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public List<Certificate> findBy(EntityFinder<Certificate> entityFinder) throws ServiceException {
+        try {
+            return dao.findBy(entityFinder);
+        } catch (DAOSQLException e) {
+            throw new ServiceException(e);
+        }
+    }
+
+    @Override
+    public List<Certificate> find(Map<String, String> params) throws ServiceException {
+        finder.newFinder();
```

Comment:
review initialisation for finder.

Reply:
I'd prefer finder creation every search query, as it was before, will it be ok?